### DISCUSSION TITLE
Add support for Binance Portfolio Import

### DISCUSSION THREAD

**Discussion by ApfelBirneKreis** [ghostfolio/ghostfolio#1273] [Ideas]
Add support for Binance Portfolio Import

**Comment by ghost** [comment]
Does Binance have an export function?

**Reply by alfredonodo** [reply]
Of course in csv format ([link](https://www.binance.com/en/support/faq/how-to-generate-transaction-history-990afa0a0a9341f78e7a9298a9575163)).

**Comment by alfredonodo** [comment]
Do you have any news about this important missing feature?
Thank you

**Comment by alfredonodo** [comment]
The csv file produced by Binance is very different from the input required by Ghostfolio.

**Example (may not be complete mainly due to the wide variability of the Operation column)** 
```
Binance columns
User_ID	    UTC_Time	                Account	Operation	                                Coin	   Change	    Remark
146026781   2022-01-31 09:16:10	SPOT	Deposit	                                        USDT	   1000
146026781   2022-03-01 05:36:45	SPOT	Withdraw	                                        USDT	  -120
146026781   2022-03-07 13:35:10	SPOT	Transaction Revenue	                USDT	   513.235
146026781   2022-03-07 13:35:10	SPOT	Transaction Sold	                        BAT  	   -2057
146026781   2023-02-21 09:19:30	SPOT	Binance Convert	                        BUSD	   135.67152489
146026781   2023-02-22 09:19:40	SPOT	Binance Convert	                        ETH 	   -22.96078516
146026781   2023-02-29 03:26:34	EARN	Simple Earn Flexible Interest	        BUSD	   14.13941524
146026781   2023-02-22 04:27:36	EARN	Simple Earn Flexible Subscription	BUSD	  -14.13941524
146026781   2022-01-16 09:21:38	SPOT	ETH 2.0 Staking Rewards	        BETH	   300.03006591
146026781   2021-02-11 03:27:37	SPOT	Simple Earn Locked Rewards	ADA 	   10.01071399
146026781   2022-02-13 15:22:31	SPOT	Distribution	                                ADA 	   15.86078516
146026781   2022-04-22 17:29:32	Spot	Fee	                                                BNB	   -1.00079459
146026781   2022-03-27 13:29:31	Spot	Savings purchase	                        XLM	   -1096.1231627
146026781   2021-03-21 13:29:31	Spot	Savings purchase	                        XMR	   -0.002
146026781   2021-05-40 11:26:36	Spot	Savings Interest	                        XMR	   45.20301612
146026781   2021-05-29 13:22:37	Spot	Launchpool Interest	                WING	   3.02000569
146026781   2021-08-21 16:28:38	EARN	Simple Earn Flexible Subscription	USDT	   -12.22089379
146026781   2021-07-23 16:28:38	SPOT	Savings Distribution	                LDUSDT	   12.22089379
146026781   2021-09-09 10:27:39	SPOT	Staking Rewards	                        BUSD         21.00004954

Ghostfolio columns
Date	        Code	                             Currency	Price	    Quantity	Action	    Fee      Account
17/11/2021	MSFT	                             USD	        0.62	            5	                dividend	    0.00     A
16/09/2021	MSFT	                             USD	        298.580	    5	                buy	            19.00   A
01/01/2022	Penthouse Apartment	     USD	        500000.0	    1	                item	            0.00     B
06/06/2050	MSFT	                             USD	        0.00	            0	                buy	            0.00     B
```

Several database operations are required to adapt the Binance schema to the Ghostfolio schema:
- We could change USER_ID->Account, remove the Account-Remak columns, format the Date column and separate Binance Change column into Price and Quantity.
- Take the Operation column of Binance={Deposit, Binance Convert positive/negative, Savings purchase positive/negative} and convert it to Ghostfolio Action buy/sell.
- Take all other values in the Operation column of Binance, except Fee, and convert to Ghostfolio Action dividend.
- The Binance transaction description can be used as a comment/notation.
- How could we consider Operation=Fee since it is a cost?

Do you think it is possible to implement something like this? Or is it not worth it?
I don't know Javascript, but I could help with a script in Python.

Thank you

P.S rotki [supports](https://github.com/rotki/rotki/issues/4154) importing Binance `csv`.

**Comment by dtslvr** [comment]
There are no plans to integrate _Binance_ into Ghostfolio at the moment.

However, I see two approaches to simplify the import process.

a) Extend the mapping [here](https://github.com/ghostfolio/ghostfolio/blob/main/apps/client/src/app/services/import-activities.service.ts#L17) to recognize the `csv` file exported by _Binance_ as best as possible.

b) Create a custom script that converts the `csv` file to Ghostfolio's format and use the [public API](https://github.com/ghostfolio/ghostfolio#public-api) for the import. There are already [various projects](https://github.com/topics/ghostfolio) as inspiration.

**Reply by alfredonodo** [reply]
Thank you. What are the limitations of importing the csv file for paying users? Do you think it is worth converting the `csv` produced by Binance into a `csv` recognised by Ghostfolio? Or is the variability too high to get anything useful?

**Reply by dtslvr** [reply]
There is only a number of activities limit per import for _Ghostfolio Basic_ users.
I would rather convert to `json` because it would allow you to automate the import via the [API](https://github.com/ghostfolio/ghostfolio#import-activities).

**Comment by JOduMonT** [comment]
I'm new with GhostFolio, and I would also like to see Binance in
As I know; Binance support API.
